internal/core/src/common/Utils.h

```diff
@@ -213,4 +214,67 @@ GetCommonPrefix(const std::string& str1, const std::string& str2) {
     return str1.substr(0, i);
 }
 
+// Iterable is a list of milvus::proto::schema::SparseFloatRow. This helper
+// function converts sparse matrix in proto to a vector of
+// knowhere::sparse::SparseRow<float>.
+template <typename Iterable>
+std::unique_ptr<knowhere::sparse::SparseRow<float>[]>
+SparseProtoToRows(const Iterable& rows) {
+    AssertInfo(rows.size() > 0, "at least 1 sparse row should be provided");
+    auto res =
+        std::make_unique<knowhere::sparse::SparseRow<float>[]>(rows.size());
+    for (size_t i = 0; i < rows.size(); ++i) {
+        knowhere::sparse::SparseRow<float> row(rows[i].indices().data_size());
+        for (int j = 0; j < rows[i].indices().data_size(); ++j) {
+            row.set_at(j, rows[i].indices().data(j), rows[i].values().data(j));
+        }
+        res[i] = std::move(row);
+    }
+    return res;
+}
+
+inline void
+SparseRowsToProto(const knowhere::sparse::SparseRow<float>* source,
+                  int64_t rows,
+                  milvus::proto::schema::SparseFloatArray* proto) {
+    int64_t max_dim = 0;
+    for (size_t i = 0; i < rows; ++i) {
+        auto new_row = proto->add_contents();
+        if (source + i == nullptr) {
+            // empty row
+            continue;
+        }
+        auto& row = source[i];
+        max_dim = std::max(max_dim, row.dim());
+        for (size_t j = 0; j < row.size(); ++j) {
+            auto [index, value] = row[j];
+            new_row->mutable_indices()->add_data(index);
+            new_row->mutable_values()->add_data(value);
+        }
+    }
+    proto->set_dim(max_dim);
+}
+
+inline int64_t
+SparseRowSerializedSize(const knowhere::sparse::SparseRow<float>& row) {
+    return row.size() * knowhere::sparse::SparseRow<float>::element_size();
+}
+
+// serialize a sparse row to bytes that will be written into binlog file.
+inline std::vector<uint8_t>
+SerializeSparseRow(const knowhere::sparse::SparseRow<float>& row) {
+    std::vector<uint8_t> buffer(SparseRowSerializedSize(row));
```

Comment:
For all in memory format, let's use knowhere::sparse::SparseRow

Other that, for all Serdes, let's use protobuf

Reply:
discussed offline: we'll use densely packed `idx, val, idx, val, ...` byte array everywhere: in `schema.proto::SparseFloatArray`, in golang/c++ memory, in persisted binlog, in placeholder of search requests, etc.

places we expect to see different format:

* in SDK and import utils we accept various formats as insert/search input, but we convert those before sending to milvus
* in restful endpoints we accept string representation like `{"indices": [102, 16, 18, ...], "values": [0.21, 0.11, 0.15, ...]}`

Reply:
the design lgtm